engine/src/stack.hpp

```diff
@@ -175,13 +178,17 @@ struct LoggingConf : public Confable {
     };
   }
 
-  void validate(const Conf& c) const override {
+  bool validate(const Conf& c, std::optional<SchemaError>& err) const override {
```

Comment:
That is quite nice, I must admit. A bit much magic for my expected users perhaps though...

The main thing is that `validate()` used to throw an exception. Since I'm making a breaking change by not throwing anymore, I want users to explicitely use one or the other in their new code.

Either `validate_or_throw` or `validate` with this SchemaError.

So I'm a little on the fence with the ValidationResult suggestion, but I definitely think the approach has its merits.

Reply:
I think not throwing exceptions for control flow is great! My suggestion mainly stems from my strive toward minimalism - having both a bool retval and an optional error ref arg is on some level redundant; especially since it's a mandatory argument. You could also think about having both, a `bool(const Conf&)` and a `bool(const Conf&, std::optional<SchemaError>&)` method; overload resolution takes care of the rest.

Reply:
I think I started out returning an `optional<SchemaError>` but there was something wrong with that approach. I can't remember anymore what it was.

I'm going to postpone this decision to a later point though, I think. The method with SchemaError is mainly intended to be used by maintainers and not so much users.